### ISSUE TITLE
Export RouteContext from library (v6)

### ISSUE THREAD

**Issue by ajsmth** [ReactTraining/react-router#7164]
It would be nice to gain access to the RouteContext and build some of our own abstractions similar to the new `Routes` component -- I would prefer using this package over a fork!

**Comment by alflennik** [comment]
I have this use case as well. I have a UI library that is router-agnostic and with several components that support plugging into any router. With v5 I can easily use the context to know if React-Router is available.

**Comment by mjackson** [comment]
I've always considered whatever is on context to be our private API, so just a fair warning that you're going to have to have a very compelling use case if you'd like us to export it. 😅 What's your use case @ajsmth?

What are you using `useIsInRouter` for @delucis? Shouldn't your components be able to be used with or without the router?

Our context API was never public in v5 @alflennik, that's what the double `__` was for in `__RouterContext`. It's always been for internal use only.

**Comment by ajsmth** [comment]
hey there - it would open up other libraries to build their own abstractions similar to useRoutes. i understand the idea behind making it private, i suppose the motivation is that it’s easier to promote a library that builds off of react router internals rather than a new implementation of a router.

**Comment by delucis** [comment]
@mjackson In my case, if the component is used within a router, it should display some navigation controls and `useNavigate`, but outside a router it only displays some controls for the current view. For now, I’ve hacked together some alternative logic, but the solution would be cleaner and less brittle if there was a way to check the router context directly.

(I feel like best practices would suggest composing things better and having the router-dependent component in some place where the router is certain to be present, but with context and hooks this kind of deep-nested logic seems more common.)

**Comment by mjackson** [comment]
@ajsmth Can you please say more about why you want to build your own `useRoutes`?

@delucis It sounds like a simple hook that returns a boolean would work in your case? Like your `useIsInRouter` hook?

**Comment by delucis** [comment]
@mjackson For my case, that would be perfect 👍

**Comment by rcarton** [comment]
I have a somewhat valid use case - we have this outdated library that uses homemade Portals (react-toolbox, that we forked a while back for accessibility) to make Dialogs and pop-up components. The problem is that `children` no longer receive the context, so we have to provide it, something like:

```
<MyContext.Consumer>
  {
    context => (
      <Dialog>
        <MyContext.Provider value={context}>
          {children}
        </MyContext>
      </Dialog>
    )
  }
</MyContext.Consumer>
```

I haven't pinpointed how/why it breaks now, but until I can update this library to not break the React API, my workaround is to "bridge" the outer component and the children by forward the contexts. I've had to do it for pretty much every library we use that uses contexts (redux, apollo, i18n, ..).

**Comment by mjackson** [comment]
@rcarton Thanks for explaining your use case. Unfortunately we can't support custom portal implementations. If I were you, I'd try to update to `React.createPortal` if possible, which should take care of the context forwarding issue.

**Comment by rcarton** [comment]
Sounds good, in the meantime I'm using `__RouterContext` as a workaround (knowing that if it breaks, it's my fault for doing BadThings™).

**Comment by ajsmth** [comment]
> @ajsmth Can you please say more about why you want to build your own `useRoutes`?
> 
> @delucis It sounds like a simple hook that returns a boolean would work in your case? Like your `useIsInRouter` hook?

@mjackson sure - from what I can tell useRoutes only renders the highest ranking route. Sometimes in a native application we would want to hold onto previously rendered routes to preserve there implicit UI (things like scroll position), I'm not sure how I would do that without modifying / customizing useRoutes

**Comment by michaelx** [comment]
Same scenario as @alflennik, UI library that is router-agnostic and on v4. The private `__RouterContex` API was a workaround, would be good to have it in v6.

**Comment by dwelle** [comment]
For future readers, the request is now being tracked here: https://github.com/ReactTraining/react-router/issues/7545

**Comment by jgoux** [comment]
Sorry I found this issue a little late.
I exposed my use-case here : https://github.com/ReactTraining/react-router/issues/7933
tl;dr : I need the access `RouteContext` in order to forward the routing context to preserve relative links when using a Slot/Fill approach.
